File: Java/src/AOA/Sort/MergeSort.java
package AOA.Sort;


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class MergeSort {
    void divide(int ar[], int l, int h) {
        int middle;
        if (l < h) {
            middle = (l + h) / 2;
            divide(ar, l, middle);
            divide(ar, middle + 1, h);
            merge(ar, l, h, middle);
        }
    }

    void merge(int ar[], int start, int end, int middle) {
        int sizeL = middle - start + 1, sizeR = end - middle;
        int[] L = new int[sizeL];
        int[] R = new int[sizeR];
        for (int i = 0; i < sizeL; ++i)
            L[i] = ar[start + i];
        for (int j = 0; j < sizeR; ++j)
            R[j] = ar[middle + j + 1];

        int i = 0, j = 0;

        // Initial index of merged subarry array
        int k = start;
        while (i < sizeL && j < sizeR) {
            if (L[i] < R[j]) ar[k++] = L[i++];
            else ar[k++] = R[j++];
        }

        while (i < sizeL) ar[k++] = L[i++];

        while (j < sizeR) ar[k++] = R[j++];


    }

    public static void main(String[] args) throws IOException {
        MergeSort ob = new MergeSort();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int arr[] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        ob.divide(arr, 0, arr.length - 1);
        for (int ar : arr) {
            System.out.print(ar + " ");
        }
    }
}

File: Python/AOA/sort/merge_sort.py
n = [4, 3, 2, 1]


def merge_sort(nums: list[int], start, end):
    if (end == start):
        return nums
    middle = (start + end) // 2
    merge_sort(nums, start, middle)
    merge_sort(nums, middle + 1, end)

    merge(nums, start, middle, end)
    return nums


def merge(nums, start, middle, end):
    L = nums[start: middle + 1]
    R = nums[middle + 1: end + 1]

    i, j, k = 0, 0, start

    while i < len(L) and j < len(R):
        if L[i] <= R[j]:
            nums[k] = L[i]
            i += 1
        else:
            nums[k] = R[j]
            j += 1
        k += 1

    while i < len(L):
        nums[k] = L[i]
        k, i = k+1, i+1
    while j < len(R):
        nums[k] = R[j]
        k, j = k+1, j+1


print(merge_sort(n, 0, len(n)))
